File: packages/shared/src/domain/execution.ts
import { ExecutionEvidence } from "./execution-evidence.js";
import { ExecutionMetadata } from "./execution-metadata.js";

/**
 * Parmana Trust Core
 *
 * Execution
 *
 * Records what actually happened after a Business
 * Transaction was authorized.
 *
 * Executions are immutable trust artifacts.
 */
export interface Execution {
  /**
   * Unique Execution identifier.
   */
  readonly executionId: string;

  /**
   * Business Transaction to which this Execution belongs.
   */
  readonly businessTransactionId: string;

  /**
   * Execution lifecycle state.
   */
  readonly status: ExecutionStatus;

  /**
   * Execution mode.
   */
  readonly mode: ExecutionMode;

  /**
   * UTC timestamp when execution started.
   */
  readonly startedAt: Date;

  /**
   * UTC timestamp when execution completed.
   *
   * Present only for terminal executions.
   */
  readonly completedAt?: Date;

  /**
   * Immutable execution evidence.
   */
  readonly evidence?: ExecutionEvidence;

  /**
   * Execution-specific metadata.
   */
  readonly metadata?: ExecutionMetadata;
}

/**
 * Canonical execution lifecycle.
 */
export enum ExecutionStatus {
  PROCESSING = "PROCESSING",

  COMPLETED = "COMPLETED",

  FAILED = "FAILED",
}

/**
 * Execution mode.
 */
export enum ExecutionMode {
  SYNC = "SYNC",

  ASYNC = "ASYNC",
}

File: python-old/parmana/models/execution.py
"""
Execution model.
"""

from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any

from .decision import Decision
from .execution_evidence import ExecutionEvidence


class ExecutionStatus(str, Enum):
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


class ExecutionMode(str, Enum):
    SYNC = "SYNC"
    ASYNC = "ASYNC"


@dataclass(frozen=True, slots=True)
class Execution:
    """
    Immutable execution artifact.
    """

    execution_id: str

    business_transaction_id: str

    decision: Decision

    status: ExecutionStatus

    mode: ExecutionMode

    started_at: datetime

    completed_at: datetime | None = None

    evidence: ExecutionEvidence | None = None

    metadata: dict[str, Any] | None = None